### PR TITLE
xds: Priority failover

### PR DESCRIPTION
**Pull request by dapengzhang0**
- Implementing [priority failover](https://docs.google.com/document/d/1fAfdllHJ3JaaKVIdrrBaMnZSHxGAnlx4BhABMQQOn_4/view?pli=1#heading=h.krx7r69l2drb)
- Removed `localityWeight` field from `LocalityLbInfo` because this is always available and up to date in `edsResponsLocalityInfo`, otherwise need to create a new instance of `LocalityLbInfo` each time weight is updated.
- Introduced `PriorityManager` that manages all the priority failover logic.

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 1]
```diff
@@ -497,5 +443,177 @@ protected Helper delegate() {
...
+          failOver();
+          return;
+        }
+
+        for (int p = 0; p < priorityTable.size(); p++) {
+          updatePriorityState(p);
+        }
+      }
+
+      private void addLocality(int priority, XdsLocality locality) {
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: Inline this as well.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 2]
```diff
@@ -91,9 +90,9 @@
     private final LoadStatsStore loadStatsStore;
     private final OrcaPerRequestUtil orcaPerRequestUtil;
     private final OrcaOobUtil orcaOobUtil;
-
-    private final Map<XdsLocality, LocalityLbInfo> localityMap = new LinkedHashMap<>();
-    private Map<XdsLocality, LocalityInfo> edsResponsLocalityInfo = ImmutableMap.of();
+    private final PriorityManager priorityManager = new PriorityManager();
+    private final Map<XdsLocality, LocalityLbInfo> localityMap = new HashMap<>();
+    private Set<XdsLocality> loadStatsStoreLocalities = ImmutableSet.of();
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: This name is wired tho. `effectiveLocalities`, `currentLocalities` or just `localities` sounds bette. And put a comment 
```java
// Localities to be load balanced instructed by traffic director.
```
would make this much more elegant.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java (hunk 1 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Done.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java (hunk 2 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Done. `effectiveLocalities` is misleading, there's multiple kinds of "effective" localities for stats store and priority manager. The latter two names sounds better.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 3]
```diff
@@ -91,8 +89,8 @@
     private final LoadStatsStore loadStatsStore;
     private final OrcaPerRequestUtil orcaPerRequestUtil;
     private final OrcaOobUtil orcaOobUtil;
-
-    private final Map<XdsLocality, LocalityLbInfo> localityMap = new LinkedHashMap<>();
+    private final PriorityManager priorityManager = new PriorityManager();
+    private final Map<XdsLocality, LocalityLbInfo> localityMap = new HashMap<>();
     private Map<XdsLocality, LocalityInfo> edsResponsLocalityInfo = ImmutableMap.of();
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: You are mutating the field, by just swapping out the immutable object (this is awkward). That is, you are changing the state of enclosing instance. This is private, implementer knows what they are doing and we should not take advantage of immutable object to save effort and avoid human mistakes.

But, this is just my opinion, you can keep it if this is your preferred way.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 4]
```diff
@@ -287,7 +244,13 @@ public void run() {
       });
 
       edsResponsLocalityInfo = ImmutableMap.copyOf(localityInfoMap);
-      onChildStateUpdated();
+      priorityManager.updateLocalityStore();
+
+      for (XdsLocality oldLocality : localityMap.keySet()) {
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Fine.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 5]
```diff
@@ -497,5 +443,174 @@ protected Helper delegate() {
...
+        }
+
+        if (priority == currentPriority && overallState != READY) {
+          failOver();
+        } else if (priority <= currentPriority && overallState == READY) {
+          goToPriority(priority);
+        }
+      }
+
+      void onLocalityStateUpdate(XdsLocality locality) {
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: A slightly better approach is to introduce a method in `PriorityManager`:

```java
// Returns the priority of the given locality or -1 if it is untracked.
int getPriority(XdsLocality locality);
```

Then you don't need to have this `PriorityManager#onLocalityStateUpdate(...)`, which is a bit awkward to be there.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: It is awkward for ChildHelper in another way:
- You would call `priorityManager` twice, `getPriority()` and `updatePriorityState(priority)`.
- You would check a magic number
  ```
    if (priority >= 0) {
      ...
    }
  ```

File: xds/src/test/java/io/grpc/xds/LocalityStoreTest.java [hunk 6]
```diff
@@ -842,6 +865,393 @@ public void reset() {
...
+        localityInfo4);
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        localityStore.updateLocalityStore(localityInfoMap);
+      }
+    });
+    assertThat(loadBalancers.keySet()).containsExactly("sz1");
+    LoadBalancer lb1 = loadBalancers.get("sz1");
+    InOrder inOrder = inOrder(lb1, helper);
```
Comment (voidzcy) [inline xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: You don't really need `inOrder`. `inOrder` verifies the sequence of interactions on one or a set of mocks, but here you are not consistently verifying the invocation sequence but keep recreating new `inOrder`s. This usage is not proper. You want to do a step-by-step check that how each child balancer helper is invoked, but you are already verifying that they are spawned one by one. So, do use `inOrder`.

Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: For each new `inOrder`, I'm consistently verifying the invocation order, until it's replaced. I need to verify the order of `helper.updateBalancingState` and `lb.handleResolvedAddressGroup`

Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: No, the order of those two methods being invoked does not matter in terms of behavior.

File: xds/src/test/java/io/grpc/xds/LocalityStoreTest.java [hunk 7]
```diff
@@ -842,6 +865,393 @@ public void reset() {
...
+    assertThat(failOverTask.isCancelled()).isFalse();
+    assertThat(fakeClock.getPendingTasks(failOverTaskFilter)).doesNotContain(failOverTask);
+    assertThat(fakeClock.getPendingTasks(failOverTaskFilter)).hasSize(1);
+    failOverTask = Iterables.getOnlyElement(fakeClock.getPendingTasks(failOverTaskFilter));
+    assertThat(fakeClock.getPendingTasks(deactivationTaskFilter)).isEmpty();
+    assertThat(loadBalancers.keySet()).containsExactly("sz1", "sz2");
+    LoadBalancer lb2 = loadBalancers.get("sz2");
+    inOrder = inOrder(lb1, lb2, helper);
+    inOrder.verify(helper).updateBalancingState(CONNECTING, BUFFER_PICKER);
+    ArgumentCaptor<ResolvedAddresses> resolvedAddressesCaptor2 =
```
Comment (voidzcy) [inline xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: nit: if you prefer using `ArgumentCaptor.forClass(null)`, use it everywhere, but not a mix of both styles.

File: xds/src/test/java/io/grpc/xds/LocalityStoreTest.java (hunk 6 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: The order is intended behavior, because I have 
```
          // In extreme case handleResolvedAddresses() may trigger updateBalancingState()
          // immediately, so execute handleResolvedAddresses() after all the setup in this method is
          // complete.
          helper.getSynchronizationContext().execute(new Runnable() {
            @Override
            public void run() {
              localityLbInfo.childBalancer.handleResolvedAddresses(
                  ResolvedAddresses.newBuilder().setAddresses(localityInfo.eags).build());
            }
          });
```

Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: That's implementation detail, i.e., __how__ you update backend addresses. The behavior you see is _backend addresses get updated_. You don't insist on the ordering of  `helper.updateBalancingState` and `lb.handleResolvedAddressGroup` as long as they are doing the desired thing.

File: xds/src/test/java/io/grpc/xds/LocalityStoreTest.java (hunk 7 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: It's already mixed in this class :(

File: xds/src/test/java/io/grpc/xds/LocalityStoreTest.java (hunk 6 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: No, I insist `updateBalancingState` triggered by `PriorityManager` happen first, because I don't want `handleResolvedAddressGroup` happen first and trigger its own `updateBalancingState` and picker in a half-baked state.

Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: Here you want to verify what happens when failing over to a uninitialized locality. You want to verify the state of that locality becomes `CONNECTING`, not even bother to verify that its backend addresses are updated first. Because it's your implementation choose to not have an unused locality holding backend addresses. It's also not wrong if you each locality holds backend addresses right after locality info is updated. Verifying the strict ordering of how these methods are called is dangerous, as you are writing tests by simulating how your implementation flows. This doesn't test anything, but just runs the implementation in test.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 8]
```diff
@@ -497,5 +443,177 @@ protected Helper delegate() {
         return orcaReportingHelperWrapper.asHelper();
       }
     }
+
+    private final class PriorityManager {
+
+      private static final int INVALID_PRIORITY = -1;
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: You don't really need this, `-1` suffices, as you are not exporting this information outside.

File: xds/src/test/java/io/grpc/xds/LocalityStoreTest.java (hunk 6 above)
Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: This is the only way I can verify the order I intended to verify. I did not use inOrder for all verifications. For any verification that inOrder is not needed I just verified without inOrder. Unfortunately there are some unintended ordering are verified, like the ordering between the following
```java
    inOrder.verify(lb1).handleResolvedAddresses(resolvedAddressesCaptor1.capture());
    assertThat(resolvedAddressesCaptor1.getValue().getAddresses()).containsExactly(eag12);
    inOrder.verify(lb2).handleResolvedAddresses(resolvedAddressesCaptor2.capture());
    assertThat(resolvedAddressesCaptor2.getValue().getAddresses()).containsExactly(eag22);
    inOrder.verify(lb3).handleResolvedAddresses(resolvedAddressesCaptor3.capture());
    assertThat(resolvedAddressesCaptor3.getValue().getAddresses()).containsExactly(eag32);
    inOrder.verify(lb4).handleResolvedAddresses(resolvedAddressesCaptor4.capture());
    assertThat(resolvedAddressesCaptor4.getValue().getAddresses()).containsExactly(eag42);
```
but it's natural deterministic ordering (It's unlikely to change when implementation is changed). All unintended ordering verification is like this. I didn't simulate my implementation flow. I just followed the test case specification and verify all the expected behavior in the specification. If implementation changes and the unintended order verification breaks, we can change the test as long as it verifies all the test spec.

Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/LocalityStoreTest.java]: What's your test specification for this test case? Precisely.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java (hunk 8 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Done.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 9]
```diff
@@ -497,5 +443,175 @@ protected Helper delegate() {
...
+        priorityTable.clear();
+        for (XdsLocality newLocality : localityInfoMap.keySet()) {
+          int priority = localityInfoMap.get(newLocality).priority;
+          while (priorityTable.size() <= priority) {
+            priorityTable.add(new ArrayList<XdsLocality>());
+          }
+          priorityTable.get(priority).add(newLocality);
+        }
+
+        if (currentPriority >= priorityTable.size()) {
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: This is verbose and misleading. You always want to run `updatePriorityState` to find the new smallest priority upon localities are refreshed. All the following lines can be reduced to 

```java
currentPriority = priorityTable.size() - 1;
for (int p = 0; i < priorityTable.size(); p++) {
  updatePriorityState(p);
}
```

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: I'm not disagreeing the logic can be simplified.
But `currentPriority` is changed in two separate methods, in your proposed change, in this method, `currentPriority` is not at a meaningful state. 
I want to clearly see `currentPriority` be in a meaningful state in every single method. Otherwise it's misleading.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 10]
```diff
@@ -497,5 +443,175 @@ protected Helper delegate() {
         return orcaReportingHelperWrapper.asHelper();
       }
     }
+
+    private final class PriorityManager {
+
+      private final List<List<XdsLocality>> priorityTable = new ArrayList<>();
+      private Map<XdsLocality, LocalityInfo> localityInfoMap = ImmutableMap.of();
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Not to make a copy now. `localityInfoMap` is still declared as `Map`. Assigning a `ImmuatbleMap` to it is only implementation detail. (My preference to use Immutables is to always have a clear snapshot view of the state.)

File: xds/src/main/java/io/grpc/xds/LocalityStore.java (hunk 9 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]:
```java
if (currentPriority >= priorityTable.size()) {
  currentPriority = priorityTable.size() - 1;
}
```
is not meaningful, at least for me.

Set `currentPriority` to `priorityTable.size() - 1` and then search for the available locality with smallest priority. Why is this not meaningful? The idea is the same as setting the upper bound to `Integer.MAX_VALUE` and then do a binary search.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Why is the 'if then' not meaningful for you? If the priorotyTable size becomes smaller - even smaller than the current priority, then you obviously have to decrease the priority value. It's obvious even you don't know about the algorithm detail.

Setting an upper bound is not obvious to me, you have to know the algorithm detail, and have to know that after a loop of the separate method `updatePriorityState()` it will change `currentPriority` to a value that makes sense, and that the loop of  `updatePriorityState()` depends an upper bound. If `updatePriorityState()`, `gotoPriority()` and `failOver()` were all inlined in this method, I could barely accept this, because then I could see clearly how the field `currentPriority` is changed in one block.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Anyway, setting `currentPriority` here is temporary, its eventual value is set after iterating over the new `priorityTable` to find the available locality with smallest priority. 

Setting `currentPriority` to `priorityTable.size() - 1` only when it exceeds the table size (use the old value if not). You have two different cases for setting a temporary value, this is not intuitive.

I am ok if you want to keep this.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Actually the upper bound logic is wrong. If changed to the upper bound logic, it will do the wrong thing at least for the initial update:

```
event detail - expected state
EDS update: P0 sz1; P1 sz2; P2 sz3; P3 sz4 - P0 C, P1 N/A, P2 N/A, P3 N/A
```

It will change `currentPriority` to 3, and the updated state will be `P0 C, P1 C, P2 C, P3 C`
```
io.grpc.xds.LocalityStoreTest > multipriority FAILED
    unexpected (3): sz2, sz4, sz3
    ---
    expected      : [sz1]
    but was       : [sz2, sz1, sz4, sz3]
        at io.grpc.xds.LocalityStoreTest.multipriority(LocalityStoreTest.java:912)
```

The algorithm is NOT simply "reset an upper bound first, and then search for the available locality with smallest priority".

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Try
```java
currentPriority = -1;
failOver();
```

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: It breaks a lot of exiting single priority tests, not only `multipriority()` test. And I quite don't follow this logic.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: I tried with 

```java
void updateLocalities(Map<XdsLocality, LocalityInfo> localityInfoMap) {
        this.localityInfoMap = ImmutableMap.copyOf(localityInfoMap);
        priorityTable.clear();
        for (XdsLocality newLocality : localityInfoMap.keySet()) {
          int priority = localityInfoMap.get(newLocality).priority;
          while (priorityTable.size() <= priority) {
            priorityTable.add(new ArrayList<XdsLocality>());
          }
          priorityTable.get(priority).add(newLocality);
        }

        cancelFailOverTimer();
        currentPriority = -1;
        failOver();
}
```

I don't see any tests break.

All you need upon locality update is to find the new `currentPriority`. Isn't this the goal of the algorithm?

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: No. It shouldn't `cancelFailOverTimer()` regardless. In some case the timer should remain.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: I should add a test case like

9 sec passes, EDS update without priority change, then after 1 sec timer should expire.

Thanks for catching a corner case.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: What's the problem for always researching the new available locality with smallest priority? That's the spec as we discussed previously.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: I think handling this corner case is flexible, we could just recompute `currentPriority` every time instead of handle the case when updated localities are exactly the same as previous ones.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: > What's the problem for always researching the new available locality with smallest priority?

There's no problem with that. But if there is no READY locality in the new update, we use the existing `currentPriority`; and if further that the existing `currentPriority` >=  `priorityTable.size()`, which does not make send, we should decrease it to `priorityTable.size() - 1`.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Upon receiving new locality update, `priorityTable` has been updated with new localities, `currentPriority` is no long valid as it points to a new set of localities. It needs to be recomputed anyway.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Let me think if we can ignore the corner case, because that is a really rare case and does not matter much. And then the code will be simplified a lot as you suggested.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: So suppose we apply the simple code path: In the corner case, i.e. in case the existing current priority is still CONNECTING, and before the 10 sec timer expires, if a new EDS update comes without priority change (e.g. only endpoint change), then it will cancel the current 10 sec timer and eagerly try the next priority immediately. Seems not big of a deal.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: > it points to a new set of localities. It needs to be recomputed anyway.

PS: This is also a minor edge case.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: > So suppose we apply the simple code path: In the corner case, i.e. in case the existing current priority is still CONNECTING, and before the 10 sec timer expires, if a new EDS update comes without priority change (e.g. only endpoint change), then it will cancel the current 10 sec timer and eagerly try the next priority immediately. Seems not big of a deal.

Why not just recomputes `currentPriority`? This can be generalized to all cases.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: In any way it affects only corner case, which we can ignore. I changed the code to reset `currentPriority` and recompute `currentPriority`.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 11]
```diff
@@ -508,17 +498,20 @@ void updatePriorityState(int priority) {
           }
         }
 
-        if (priority == currentPriority || overallState == READY) {
+        if (priority == currentPriority) {
+          updatePicker(overallState, childPickers);
+          if (overallState == READY) {
+            cancelFailOverTimer();
+            goToPriority(priority);
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: This is not necessary.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: This is necessary, because `currentPriority` after the change is a temporary value. That's why I didn't like `currentPriority`  to be temporary value cross multiple methods.

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Now inlined `goToPriority()`.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 12]
```diff
@@ -497,5 +443,164 @@ protected Helper delegate() {
...
+          }
+        }
+
+        if (priority == currentPriority) {
+          updatePicker(overallState, childPickers);
+          if (overallState == READY) {
+            cancelFailOverTimer();
+          } else if (overallState == TRANSIENT_FAILURE) {
+            cancelFailOverTimer();
+            failOver();
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: This two lines together is confusing. The only case you need to cancel the old failover and restart a new failover with current priority is when recomputing `currentPriority` from `updateLocalities(...)`. So why not do 

```java
cancelFailoverTimer();
currentPriority = -1;
failOver();
```

as I mentioned before?

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: >The only case you need to cancel the old failover and restart a new failover with current priority is when recomputing currentPriority from updateLocalities(...).

This is not true. In the spec:

> ... the xds policy will start a timer for 10 seconds to allow the new locality map time to establish connections to its endpoints.  If the locality map has not reported state READY by the time the timer fires, or if the locality map reports state TRANSIENT_FAILURE before the timer fires, the xds policy will consider the map unreachable and will switch to a locality map for the priority N+1 localities.

This is exactly following the spec, why is it confusing? On the contrary, 

```
cancelFailoverTimer();
currentPriority = -1;
failOver();
```
Is a weird behavior/optimization that is not specified, and behaves differently from the spec on edge cases.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Oh you are right. You would need to cancel timer when getting TRANSIENT_FAILURE.

**Comment by dapengzhang0** [comment]
@voidzcy Thanks a lot for your thorough review.
